### PR TITLE
Bringup: gemma4/12B [ 8-way tensor parallel]

### PR DESCRIPTION
**Pull request by saiarthiraguram**
## Github issue
[Issue Link](https://github.com/tenstorrent/tt-xla/issues/5187)

## Problem Description
Adds the `gemma4` text-only causal-LM loader (`google/gemma-4-12B`,
`Gemma4UnifiedForConditionalGeneration`) and an 8-way tensor-parallel
inference test on `n300-llmbox`.

## Reasons
`google/gemma-4-12B` is a brand-new `gemma4_unified` any-to-any (text +
vision + audio) arch. This brings up the **text** causal-LM path. bf16
weights (~22 GiB) don't fit a single Wormhole chip → tensor-parallel on a
(1, 8) mesh (all 8 chips / 4 boards).

## Highlights
- Loader      : `tt_forge_models/gemma4/pytorch/loader.py` (submodule)
- Params      : ~11.96 B (full unified incl. vision/audio towers; text path tested)
- Modality    : multimodal arch, **text path** only
- Parallelism : tensor_parallel, TP=8, mesh (1, 8) ["batch", "model"]
- Archs       : n300-llmbox
- PCC         : passed at default threshold (0.99), CPU vs TT, real weights
- Transformers: requires `5.10.0.dev0` (git main) — brand-new arch

## Shard strategy
FSDP-style Pattern B. Column-parallel `q_proj` + MLP `gate/up`; row-parallel
`o_proj` + `down_proj`. **KV projections replicated**: the 8 global
`full_attention` layers use `attention_k_eq_v` (so `v_proj is None`, value
reuses key) and carry a single `num_global_key_value_heads=1` KV head that
cannot be sharded across the mesh — replicating all KV is the standard
GQA-TP fallback and keeps query heads correctly grouped per chip. Norms,
embeddings, lm_head and the unused vision/audio towers are replicated.

## Loader specifics (why this arch needed care)
- `use_cache` is set on the config, not passed to `from_pretrained`
  (`Gemma4UnifiedForConditionalGeneration.__init__` rejects the kwarg).
- `load_inputs` returns a **dict** `{input_ids, attention_mask}` so the
  harness binds by name — `pixel_values` is the 2nd positional param, so a
  positional list would mis-bind the attention mask and fire the vision tower.
- `unpack_forward_output` → `.logits` (custom `Gemma4UnifiedCausalLMOutputWithPast`).

## Test
`tests/runner/test_models.py::test_all_models_torch[gemma4/pytorch-12B-tensor_parallel-inference]`
— passes on n300-llmbox.

```
TT_VISIBLE_DEVICES=0,1,2,3 TT_XLA_SPMD=1 CONVERT_SHLO_TO_SHARDY=1 \
  TT_XLA_ARCH=n300-llmbox HF_HOME=/proj_sw/user_dev/ctr-sraguram/hf_cache \
  pytest -svv "tests/runner/test_models.py::test_all_models_torch[gemma4/pytorch-12B-tensor_parallel-inference]"
```

## Logs
- Steps log : 
[bringup_steps.txt](https://github.com/user-attachments/files/28853872/bringup_steps.txt)

- Overview  : 
[model_overview.md](https://github.com/user-attachments/files/28853876/model_overview.md)

- Verify log:
[iter3_verify_tp.log](https://github.com/user-attachments/files/28853877/iter3_verify_tp.log)

### CODE REVIEW

File: gemma4/pytorch/loader.py [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+        every query head correctly grouped on each chip. ``k_proj``/``v_proj``
+        are therefore skipped (and guarded for absence/None).
+
+        Per-projection RMSNorms (q_norm/k_norm/v_norm), layernorms, embeddings
+        and lm_head are left replicated. Vision/audio towers are unused on the
+        text-only path and are replicated.
+        """
+        shard_specs = {}
+        for layer in model.model.language_model.layers:
+            shard_specs[layer.mlp.gate_proj.weight] = ("model", "batch")
```
Comment (vkovinicTT) [inline gemma4/pytorch/loader.py]: When sharding instead of `"batch"`, please use `None` as that is sharding in Megatron style.

**Review by vkovinicTT — CHANGES_REQUESTED** [review]

File: gemma4/pytorch/loader.py (hunk 1 above)
Comment (saiarthiraguram) [reply gemma4/pytorch/loader.py]: Hi @vkovinicTT , made the required change in recent commit https://github.com/tenstorrent/tt-forge-models/pull/731/commits/5535e6e5f3f8a1ee2493051119a5d9d4aee93f6d
### Logs
[iter4_verify_tp.log](https://github.com/user-attachments/files/28872439/iter4_verify_tp.log)

**Review by vkovinicTT — APPROVED** [review]
